[BPF] Handle traps with kfunc call __bpf_trap

yonghong-song opened the pull request:
Currently, middle-end generates 'unreachable' insn if the compiler
feels the code is indeed unreachable or the code becomes invalid
due to some optimizaiton (e.g. code optimization with uninitialized
variables).

Right now BPF backend ignores 'unreachable' insn during selectiondag
lowering. For cases where 'unreachable' is due to invalid code
transformation, such a signal will be missed. Later on, users needs
some effort to debug it which impacts developer productivity.

This patch enabled selectiondag lowering for 'unreachable' insn.

Previous attempt ([1]) tries to have a backend IR pass to filter
out 'unreachable' insns in a number of cases. But such pattern
matching may misalign with future middle-end optimization with
'unreachable' insns.

This patch takes a different approach. The 'unreachable' insn is
lowered with special encoding in bpf object file and verifier
will do proper verification for the bpf prog. More specifically,
the 'unreachable' insn is replaced by a __bpf_trap() function.
This function will be a kfunc (in ".ksyms" section) with a weak
attribute, but does not have definition. The actual kfunc definition
is expected to be in kernel. The __bpf_trap() extern function
is also encoded in BTF. The name __bpf_trap() is chosen to satisfy
reserved identifier requirement.

Besides the uninitialized variable case, the builtin function
'__builtin_trap' can also generate kfunc __bpf_trap(). For example
in [3], we have
```
  # define __bpf_unreachable()  __builtin_trap()
```
If the compiler didn't remove __builtin_trap() during middle-end
optimization, compilation will fail.

With this patch, compilation will not fail and __builtin_trap()
is converted to __bpf_trap() kfunc. The eventual failure will be
in verifier instead of llvm compilation. To keep compilation
time failure, user can add an option like `-ftrap-function=<something>`.

I tested this patch on bpf selftests and all tests are passed.
I also tried original example in [2] and the code looks like below:
```
          ; {
               0:       bf 16 00 00 00 00 00 00 r6 = r1
          ;       bpf_printk("Start");
               1:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x0 ll
                        0000000000000008:  R_BPF_64_64  .rodata
               3:       b4 02 00 00 06 00 00 00 w2 = 0x6
               4:       85 00 00 00 06 00 00 00 call 0x6
          ; DEFINE_FUNC_CTX_POINTER(data)
               5:       61 61 4c 00 00 00 00 00 w1 = *(u32 *)(r6 + 0x4c)
          ;       bpf_printk("pre ipv6_hdrlen_offset");
               6:       18 01 00 00 06 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x6 ll
                        0000000000000030:  R_BPF_64_64  .rodata
               8:       b4 02 00 00 17 00 00 00 w2 = 0x17
               9:       85 00 00 00 06 00 00 00 call 0x6
              10:       85 10 00 00 ff ff ff ff call -0x1
                        0000000000000050:  R_BPF_64_32  __bpf_trap
              11:       95 00 00 00 00 00 00 00 exit
          <END>
```
Eventually kernel verifier will emit the following logs:
```
      10: (85) call __bpf_trap#74479
      unexpected __bpf_trap() due to uninitialized variable?
```
In another internal sched-ext bpf prog, with the patch we have bpf code:
```
  Disassembly of section .text:
  0000000000000000 <scx_storage_init_single>:
  ; {
       0:       bc 13 00 00 00 00 00 00 w3 = w1
       1:       b4 01 00 00 00 00 00 00 w1 = 0x0
  ;       const u32 zero = 0;
  ...
  0000000000003a80 <create_dom>:
  ; {
    1872:       bc 16 00 00 00 00 00 00 w6 = w1
  ;       bpf_printk("dom_id %d", dom_id);
    1873:       18 01 00 00 3f 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x3f ll
                0000000000003a88:  R_BPF_64_64  .rodata
    1875:       b4 02 00 00 0a 00 00 00 w2 = 0xa
    1876:       bc 63 00 00 00 00 00 00 w3 = w6
    1877:       85 00 00 00 06 00 00 00 call 0x6
  ;       ret = scx_bpf_create_dsq(dom_id, 0);
    1878:       bc 61 00 00 00 00 00 00 w1 = w6
    1879:       b4 02 00 00 00 00 00 00 w2 = 0x0
    1880:       85 10 00 00 ff ff ff ff call -0x1
                0000000000003ac0:  R_BPF_64_32  scx_bpf_create_dsq
  ;       domc->node_cpumask = node_data[node_id];
    1881:       85 10 00 00 ff ff ff ff call -0x1
                0000000000003ac8:  R_BPF_64_32  __bpf_trap
    1882:       95 00 00 00 00 00 00 00 exit
    <END>
```
The verifier can easily report the error too.

A bpf flag `-bpf-disable-trap-unreachable` is introduced to disable
trapping for 'unreachable' or __builtin_trap.

  [1] https://github.com/llvm/llvm-project/pull/126858
  [2] https://github.com/msune/clang_bpf/blob/main/Makefile#L3
  [3] https://github.com/libbpf/libbpf/blob/master/src/bpf_helpers.h

Commit:
[BPF] Handle traps with kfunc call __bpf_trap

NOTE: not working as the symbol is generated at selectiondag stage

Currently, middle-end generates 'unreachable' insn if the compiler
feels the code is indeed unreachable or the code becomes invalid
due to some optimizaiton (e.g. code optimization with uninitialized
variables).

Right now BPF backend ignores 'unreachable' insn during selectiondag
lowering. For cases where 'unreachable' is due to invalid code
transformation, such a signal will be missed. Later on, users needs
some effort to debug it which impacts developer productivity.

This patch enabled selectiondag lowering for 'unreachable' insn.

Previous attempt ([1]) tries to have a backend IR pass to filter
out 'unreachable' insns in a number of cases. But such pattern
matching may misalign with future middle-end optimization with
'unreachable' insns.

This patch takes a different approach. The 'unreachable' insn is
lowered with special encoding in bpf object file and verifier
will do proper verification for the bpf prog. More specifically,
the 'unreachable' insn is replaced by a __bpf_trap() function.
This function will be a kfunc (in ".ksyms" section) with a weak
attribute, but does not have definition. The actual kfunc definition
is expected to be in kernel. The __bpf_trap() extern function
is also encoded in BTF. The name __bpf_trap() is chosen to satisfy
reserved identifier requirement.

Besides the uninitialized variable case, the builtin function
'__builtin_trap' can also generate kfunc __bpf_trap(). For example
in [3], we have
  # define __bpf_unreachable()	__builtin_trap()
If the compiler didn't remove __builtin_trap() during middle-end
optimization, compilation will fail.

With this case, compilation will not fail and __builtin_trap()
is converted to __bpf_trap() kfunc. The eventual failure will be
in verifier instead of llvm compilation.

I tested this patch on bpf selftests and all tests are passed.
I also tried original example in [2] and the code looks like below:

          ; {
               0:       bf 16 00 00 00 00 00 00 r6 = r1
          ;       bpf_printk("Start");
               1:       18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x0 ll
                        0000000000000008:  R_BPF_64_64  .rodata
               3:       b4 02 00 00 06 00 00 00 w2 = 0x6
               4:       85 00 00 00 06 00 00 00 call 0x6
          ; DEFINE_FUNC_CTX_POINTER(data)
               5:       61 61 4c 00 00 00 00 00 w1 = *(u32 *)(r6 + 0x4c)
          ;       bpf_printk("pre ipv6_hdrlen_offset");
               6:       18 01 00 00 06 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x6 ll
                        0000000000000030:  R_BPF_64_64  .rodata
               8:       b4 02 00 00 17 00 00 00 w2 = 0x17
               9:       85 00 00 00 06 00 00 00 call 0x6
              10:       85 10 00 00 ff ff ff ff call -0x1
                        0000000000000050:  R_BPF_64_32  __bpf_trap
              11:       95 00 00 00 00 00 00 00 exit
          <END>

Eventually kernel verifier will emit the following logs:
      10: (85) call __bpf_trap#74479
      unexpected __bpf_trap() due to uninitialized variable?

In another internal sched-ext bpf prog, with the patch we have bpf code:
  Disassembly of section .text:
  0000000000000000 <scx_storage_init_single>:
  ; {
       0:       bc 13 00 00 00 00 00 00 w3 = w1
       1:       b4 01 00 00 00 00 00 00 w1 = 0x0
  ;       const u32 zero = 0;
  ...
  0000000000003a80 <create_dom>:
  ; {
    1872:       bc 16 00 00 00 00 00 00 w6 = w1
  ;       bpf_printk("dom_id %d", dom_id);
    1873:       18 01 00 00 3f 00 00 00 00 00 00 00 00 00 00 00 r1 = 0x3f ll
                0000000000003a88:  R_BPF_64_64  .rodata
    1875:       b4 02 00 00 0a 00 00 00 w2 = 0xa
    1876:       bc 63 00 00 00 00 00 00 w3 = w6
    1877:       85 00 00 00 06 00 00 00 call 0x6
  ;       ret = scx_bpf_create_dsq(dom_id, 0);
    1878:       bc 61 00 00 00 00 00 00 w1 = w6
    1879:       b4 02 00 00 00 00 00 00 w2 = 0x0
    1880:       85 10 00 00 ff ff ff ff call -0x1
                0000000000003ac0:  R_BPF_64_32  scx_bpf_create_dsq
  ;       domc->node_cpumask = node_data[node_id];
    1881:       85 10 00 00 ff ff ff ff call -0x1
                0000000000003ac8:  R_BPF_64_32  __bpf_trap
    1882:       95 00 00 00 00 00 00 00 exit
    <END>
The verifier can easily report the error too.

A bpf flag `-bpf-disable-trap-unreachable` is introduced to disable
trapping for 'unreachable' or __builtin_trap.

  [1] https://github.com/llvm/llvm-project/pull/126858
  [2] https://github.com/msune/clang_bpf/blob/main/Makefile#L3
  [3] https://github.com/libbpf/libbpf/blob/master/src/bpf_helpers.h

yonghong-song commented:
cc @anakryiko @jemarch

nikic commented on llvm/lib/Target/BPF/BPFISelLowering.cpp:
```diff
@@ -726,6 +733,38 @@ SDValue BPFTargetLowering::LowerATOMIC_LOAD_STORE(SDValue Op,
+  MachineFunction &MF = DAG.getMachineFunction();
+  Function &F = MF.getFunction();
+  if (F.hasFnAttribute(Attribute::Naked))
+    return Op;
```
Probably this should be part of the generic handling for TrapUnreachable. IMHO we should never insert extra code into a naked function, for other targets as well.

yonghong-song replied:
Sounds good to me. Let me check more and may have a patch for this later.

nikic reviewed (commented):
The general approach here looks reasonable to me.

How does the resulting verifier error look like?

anakryiko commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -1622,6 +1627,25 @@ void BTFDebug::endModule() {
 
+  // Create a BTF entry for func __unreachable_helper.
+  const Module *M = MMI->getModule();
+  Function *F = M->getFunction("__unreachable_helper");
```
I don't like the approach of Clang emitting kfunc call, putting it in .ksyms section (which is just a libbpf convention, not some sort of gold standard or anything), and so on. I think the appropriate way to do with would be a BPF instruction. 
But if we have to go with kfunc approach, let's at least call it `bpf_unreachable()`. We can even implement this in the kernel (it would just WARN/panic), so libbpf successfully resolves it without unnecessary warnings on modern kernels. And BPF verifier would provide a meaningful error message if it detects reachable `bpf_unreachable()` call.

yonghong-song replied:
From llvm perspective, adding a new insn is easy, likes below:
```
diff --git a/llvm/lib/Target/BPF/BPFInstrInfo.td b/llvm/lib/Target/BPF/BPFInstrInfo.td
index 2dcf1eae086b..59bad209ea54 100644
--- a/llvm/lib/Target/BPF/BPFInstrInfo.td
+++ b/llvm/lib/Target/BPF/BPFInstrInfo.td
@@ -790,11 +790,25 @@ class RET<string OpcodeStr>
   let BPFClass = BPF_JMP;
 }
 
+class TRAP<string OpcodeStr>
+    : TYPE_ALU_JMP<BPF_EXIT.Value, BPF_K.Value,
+                   (outs),
+                   (ins),
+                   !strconcat(OpcodeStr, ""),
+                   []> {
+  let Inst{31-0} = 1;
+  let BPFClass = BPF_JMP;
+}
+
 let isReturn = 1, isTerminator = 1, hasDelaySlot=0, isBarrier = 1,
     isNotDuplicable = 1 in {
   def RET : RET<"exit">;
 }
 
+let isTrap = 1 in {
+  def UNREACHABLE : TRAP<"unreachable">;
+} // isTerminator = 1
+
 // ADJCALLSTACKDOWN/UP pseudo insns
 let Defs = [R11], Uses = [R11], isCodeGenOnly = 1 in {
 def ADJCALLSTACKDOWN : Pseudo<(outs), (ins i64imm:$amt1, i64imm:$amt2),
```
Different encoding is possible. I am chosing current kfunc approach since it minimizes the kernel change, spares one new insn, and can work on old kernel.

But I am open to the new insn approach or real bpf_unreachable() kfunc approach if we can reach a consensus.

eddyz87 replied:
While `.ksyms` is just a convention, I think that emitting kfunc calls from clang might be useful. E.g. currently `memcpy`, `memset`, `memmove`, `memcmp` intrinsics are unrolled, but there is no point in doing so.

4ast replied:
> While `.ksyms` is just a convention, I think that emitting kfunc calls from clang might be useful. E.g. currently `memcpy`, `memset`, `memmove`, `memcmp` intrinsics are unrolled, but there is no point in doing so.

This convention is pretty much a de-facto standard now. Several libraries support it.
It's not unusual for a compiler to use a particular section name like ".text" or ".rodata" for specific needs.
The name is target dependent. ".ksyms" is another special name.
So I think it's appropriate for a compiler to emit "call blah_unreachable" and "call blah_memcpy" with ".ksyms" section name. It's cleaner and more flexible than "call magic_const" or inventing a new trap-like insn.

yonghong-song commented:
> How does the resulting verifier error look like?

The current verifier error looks like
```
  func#0 @0
  last insn is not an exit or jmp
```

This is due to that there is a logic in verifier to ensure the last insn in the func must be an 'exit' or 'jmp' insn. If this approach sounds reasonable, if __unreachable_helper is the last insn, the verifier log can be changed to e.g.,
    last insn is marked as unreachable, maybe due to uninitialized variable?
Such more explicit message will prompt users to check their auto variables.

Without this patch, the user will simply get error message
   last insn is not an exit or jmp
and users needs to dig into asm codes to find why the code is generated that way.

eddyz87 commented:
Maybe consider adding a test case?
E.g. as below:

```llvm
; RUN: llc -mtriple=bpfel -mcpu=v3 -filetype=obj -o %t1 %s
; RUN: llvm-objcopy --dump-section='.BTF'=%t2 %t1
; RUN: %python %p/print_btf.py %t2 | FileCheck -check-prefixes=CHECK-BTF %s
; RUN: llc -mtriple=bpfel -mcpu=v3 < %s | FileCheck -check-prefixes=CHECK %s

define void @foo() {
entry:
  tail call void @bar()
  unreachable
}

; CHECK:      foo:
; CHECK-NEXT: .Lfunc_begin0:
; CHECK-NEXT:   .cfi_startproc
; CHECK-NEXT: # %bb.0:
; CHECK-NEXT:    call bar
; CHECK-NEXT:    call __unreachable_helper
; CHECK-NEXT: .Lfunc_end0:

define void @buz() #0 {
entry:
  tail call void asm sideeffect "r0 = r1; exit;", ""()
  unreachable
}

; CHECK: buz:
; CHECK-NEXT: .Lfunc_begin1:
; CHECK-NEXT:   .cfi_startproc
; CHECK-NEXT: # %bb.0:
; CHECK-NEXT:   #APP
; CHECK-NEXT:   r0 = r1
; CHECK-NEXT:   exit
; CHECK-EMPTY:
; CHECK-NEXT:   #NO_APP
; CHECK-NEXT: .Lfunc_end1:

; CHECK-BTF: [1] FUNC_PROTO '(anon)' ret_type_id=0 vlen=0
; CHECK-BTF: [2] FUNC '__unreachable_helper' type_id=1 linkage=extern
; CHECK-BTF: [3] DATASEC '.ksyms' size=0 vlen=1
; CHECK-BTF:  type_id=2 offset=0 size=0

declare dso_local void @bar()

attributes #0 = { naked }

!llvm.dbg.cu = !{!0}
!llvm.module.flags = !{!2, !3}

!0 = distinct !DICompileUnit(language: DW_LANG_C11, file: !1, emissionKind: FullDebug)
!1 = !DIFile(filename: "test.c", directory: "/some/dir")
!2 = !{i32 7, !"Dwarf Version", i32 5}
!3 = !{i32 2, !"Debug Info Version", i32 3}
```

eddyz87 commented on llvm/lib/Target/BPF/BPFISelLowering.cpp:
```diff
@@ -726,6 +733,38 @@ SDValue BPFTargetLowering::LowerATOMIC_LOAD_STORE(SDValue Op,
+    Function *NewF = Function::Create(FT, GlobalValue::ExternalWeakLinkage,
+                                      "__unreachable_helper", M);
+    NewF->setDSOLocal(true);
+    NewF->setCallingConv(CallingConv::C);
```
Since we do not intend to execute this function, or any code after this function, maybe consider `CallingConv::PreserveAll`? Like with [fastcall](https://github.com/llvm/llvm-project/blob/432463de550f567ad782f48ac8080c03de8085d4/llvm/lib/Target/BPF/BPFISelLowering.cpp#L552).

yonghong-song replied:
I changed from CallingConv::C to CallingConv::PreserveAll and the generated asm code is the same. So looks like either is fine. Since The Function NewF has 0 argument and return void, I probably will keep CallingConv::C.

yonghong-song replied:
Let me try some experiments on this with heavy register pressure.

eddyz87 replied:
E.g. it would be good for the following not to generate too much spills:

```llvm
$ cat test6.ll
; ModuleID = 'test6.c'
source_filename = "test6.c"
target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
target triple = "bpf"

define void @foo(i1 %p) {
entry:
  %a = call i32 @bar()
  %b = call i32 @bar()
  %c = call i32 @bar()
  %d = call i32 @bar()
  %e = call i32 @bar()
  %f = call i32 @bar()
  %g = call i32 @bar()
  %h = call i32 @bar()
  %k = call i32 @bar()

  br i1 %p, label %tlabel, label %flabel

flabel:
  unreachable

tlabel:
  %r1 = add i32 0,   %a
  %r2 = add i32 %r1, %b
  %r3 = add i32 %r2, %c
  %r4 = add i32 %r3, %d
  %r5 = add i32 %r4, %e
  %r6 = add i32 %r5, %f
  %r7 = add i32 %r6, %g
  %r8 = add i32 %r7, %h
  %r9 = add i32 %r8, %k
  call void @buz(i32 %r9)
  ret void
}


declare dso_local i32 @bar()
declare dso_local void @buz(i32)

!llvm.dbg.cu = !{!0}
!llvm.module.flags = !{!2, !3}

!0 = distinct !DICompileUnit(language: DW_LANG_C11, file: !1, emissionKind: FullDebug)
!1 = !DIFile(filename: "test.c", directory: "/some/dir")
!2 = !{i32 7, !"Dwarf Version", i32 5}
!3 = !{i32 2, !"Debug Info Version", i32 3}
```
```asm
$ llc -mtriple=bpfel -mcpu=v3 < test6.ll
	.file	"test6.c"
	.text
	.globl	foo                             # -- Begin function foo
	.p2align	3
	.type	foo,@function
foo:                                    # @foo
.Lfunc_begin0:
	.cfi_startproc
# %bb.0:                                # %entry
	w7 = w1
	call bar
	w6 = w0
	call bar
	*(u32 *)(r10 - 8) = w0
	call bar
	*(u32 *)(r10 - 16) = w0
	call bar
	*(u32 *)(r10 - 24) = w0
	call bar
	*(u32 *)(r10 - 32) = w0
	call bar
	*(u32 *)(r10 - 40) = w0
	call bar
	w9 = w0
	call bar
	w8 = w0
	call bar
	w7 &= 1
	if w7 != 0 goto LBB0_2
# %bb.1:                                # %flabel
	call __unreachable_helper
LBB0_2:                                 # %tlabel
	w1 = *(u32 *)(r10 - 8)
	w6 += w1
	w1 = *(u32 *)(r10 - 16)
	w6 += w1
	w1 = *(u32 *)(r10 - 24)
	w6 += w1
	w1 = *(u32 *)(r10 - 32)
	w6 += w1
	w1 = *(u32 *)(r10 - 40)
	w6 += w1
	w6 += w9
	w6 += w8
	w6 += w0
	w1 = w6
	call buz
	exit
	...
```

yonghong-song replied:
We should be okay here. 5 stack slots are necessary due to `if w7 != 0 goto LBB0_2` going to `LBB0_2`. I tested with both calling convention C and PreserveAll. The result is the same.

eddyz87 replied:
I see, my example was broken because of calls to `bar()` that needed stack spill, sorry about that.
If changed to something like  [this](https://github.com/user-attachments/files/19399316/test6.ll.txt), there are no spills generated. But this makes me wonder how does it work:

```
$ llc -debug-only=isel -mtriple=bpfel -mcpu=v3 < test6.ll
...
*** MachineFunction at end of ISel ***
# Machine code for function foo: IsSSA, TracksLiveness
Function Live Ins: $w1 in %9, $r3 in %11

bb.0.entry:
  successors: %bb.2(0x80000000), %bb.1(0x00000000); %bb.2(100.00%), %bb.1(0.00%)
  liveins: $w1, $r3
  %11:gpr = COPY $r3
  %9:gpr32 = COPY $w1
  %0:gpr32 = LDW32 %11:gpr, 0 :: (volatile load (s32) from %ir.ap1)
  %1:gpr32 = LDW32 %11:gpr, 4 :: (volatile load (s32) from %ir.bp)
  %2:gpr32 = LDW32 %11:gpr, 8 :: (volatile load (s32) from %ir.cp)
  %3:gpr32 = LDW32 %11:gpr, 12 :: (volatile load (s32) from %ir.dp)
  %4:gpr32 = LDW32 %11:gpr, 16 :: (volatile load (s32) from %ir.ep)
  %5:gpr32 = LDW32 %11:gpr, 20 :: (volatile load (s32) from %ir.fp)
  %6:gpr32 = LDW32 %11:gpr, 24 :: (volatile load (s32) from %ir.gp)
  %7:gpr32 = LDW32 %11:gpr, 28 :: (volatile load (s32) from %ir.hp)
  %8:gpr32 = LDW32 %11:gpr, 32 :: (volatile load (s32) from %ir.kp)
  %12:gpr32 = AND_ri_32 %9:gpr32(tied-def 0), 1
  JNE_ri_32 killed %12:gpr32, 0, %bb.2
  JMP %bb.1

bb.1.flabel:
; predecessors: %bb.0

  ADJCALLSTACKDOWN 0, 0, implicit-def dead $r11, implicit $r11
  JAL &__unreachable_helper, <regmask $r6 $r7 $r8 $r9 $r10 $w6 $w7 $w8 $w9 $w10>, implicit $r11, implicit-def $r11
  ADJCALLSTACKUP 0, 0, implicit-def dead $r11, implicit $r11

bb.2.tlabel:
; predecessors: %bb.0

  %13:gpr32 = ADD_rr_32 %0:gpr32(tied-def 0), %1:gpr32
  %14:gpr32 = ADD_rr_32 %13:gpr32(tied-def 0), %2:gpr32
  %15:gpr32 = ADD_rr_32 %14:gpr32(tied-def 0), %3:gpr32
  %16:gpr32 = ADD_rr_32 %15:gpr32(tied-def 0), %4:gpr32
  %17:gpr32 = ADD_rr_32 %16:gpr32(tied-def 0), %5:gpr32
  %18:gpr32 = ADD_rr_32 %17:gpr32(tied-def 0), %6:gpr32
  %19:gpr32 = ADD_rr_32 %18:gpr32(tied-def 0), %7:gpr32
  %20:gpr32 = ADD_rr_32 %19:gpr32(tied-def 0), %8:gpr32
  ADJCALLSTACKDOWN 0, 0, implicit-def dead $r11, implicit $r11
  $w1 = COPY %20:gpr32
  JAL @buz, <regmask $r6 $r7 $r8 $r9 $r10 $w6 $w7 $w8 $w9 $w10>, implicit $r11, implicit $w1, implicit-def $r11
  ADJCALLSTACKUP 0, 0, implicit-def dead $r11, implicit $r11
  RET
...
```

Note that basic block `bb.1.flabel` does not have terminator after the `__unreachable_helper` call. Is it ok?
Is it necessary to mark to add `CLI.DoesNotReturn = true` in `LowerTRAP`?

eddyz87 replied:
> basic block bb.1.flabel does not have terminator after the __unreachable_helper call

Basing on the comments for `TargetInstrInfo::analyzeBranch` absence of terminators means falltrhough.
But if that is so, how come call to `__unreachable_helper` does not clobber r1-r5 in this example?
From llvm point of view there is a path `bb.0.entry` -> `bb.1.flabel` -> `bb.2.tlabel`, but register allocator somehow concludes that r1-r5 are intact after `bb.1.flabel`.

yonghong-song replied:
I think the real reason is that def-use chain will not be able to cross 'unreachable' insn, even after lowering. The optimization is really depending on data flow, so I think it is fine that we do not have CLI.DoesNotReturn = true for __unreachable_helper.

yonghong-song replied:
Thinking again, maybe adding CLI.DoesNotReturn = true is a good idea, just in case.

eddyz87 commented on llvm/lib/Target/BPF/BPFISelLowering.cpp:
```diff
@@ -521,10 +526,12 @@ SDValue BPFTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
-    fail(CLI.DL, DAG,
-         Twine("A call to built-in function '" + StringRef(E->getSymbol()) +
-               "' is not supported."));
+    if (strcmp(E->getSymbol(), "__unreachable_helper") != 0) {
```
Nit: maybe declare `"__unreachable_helper"` as a constant?

yonghong-song replied:
Do you mean a constant variable? This probably not a good idea as the func later is added to BTF. Or you mean a constant function? What does it mean for a constant function?

eddyz87 replied:
I mean something like `static const char *UNREACHABLE_HELPER = "__unreachable_helper";` at the file level, or a `#define ...`.

yonghong-song replied:
I see. Yes, it is a good idea. Will do this.

eddyz87 reviewed (approved):
Overall lgtm, modulo absence of tests.

yonghong-song commented:
I will add some tests in the next revision.

> Maybe consider adding a test case? E.g. as below:
> 
> ```llvm
> ; RUN: llc -mtriple=bpfel -mcpu=v3 -filetype=obj -o %t1 %s
> ; RUN: llvm-objcopy --dump-section='.BTF'=%t2 %t1
> ; RUN: %python %p/print_btf.py %t2 | FileCheck -check-prefixes=CHECK-BTF %s
> ; RUN: llc -mtriple=bpfel -mcpu=v3 < %s | FileCheck -check-prefixes=CHECK %s
> 
> define void @foo() {
> entry:
>   tail call void @bar()
>   unreachable
> }
> 
> ; CHECK:      foo:
> ; CHECK-NEXT: .Lfunc_begin0:
> ; CHECK-NEXT:   .cfi_startproc
> ; CHECK-NEXT: # %bb.0:
> ; CHECK-NEXT:    call bar
> ; CHECK-NEXT:    call __unreachable_helper
> ; CHECK-NEXT: .Lfunc_end0:
> 
> define void @buz() #0 {
> entry:
>   tail call void asm sideeffect "r0 = r1; exit;", ""()
>   unreachable
> }

As suggested by @nikic, I will later have a patch to prevent 'unreachable' for naked functions for all architectures. So this test will not be included.

> 
> ; CHECK: buz:
> ; CHECK-NEXT: .Lfunc_begin1:
> ; CHECK-NEXT:   .cfi_startproc
> ; CHECK-NEXT: # %bb.0:
> ; CHECK-NEXT:   #APP
> ; CHECK-NEXT:   r0 = r1
> ; CHECK-NEXT:   exit
> ; CHECK-EMPTY:
> ; CHECK-NEXT:   #NO_APP
> ; CHECK-NEXT: .Lfunc_end1:
> 
> ; CHECK-BTF: [1] FUNC_PROTO '(anon)' ret_type_id=0 vlen=0
> ; CHECK-BTF: [2] FUNC '__unreachable_helper' type_id=1 linkage=extern
> ; CHECK-BTF: [3] DATASEC '.ksyms' size=0 vlen=1
> ; CHECK-BTF:  type_id=2 offset=0 size=0
> 
> declare dso_local void @bar()
> 
> attributes #0 = { naked }
> 
> !llvm.dbg.cu = !{!0}
> !llvm.module.flags = !{!2, !3}
> 
> !0 = distinct !DICompileUnit(language: DW_LANG_C11, file: !1, emissionKind: FullDebug)
> !1 = !DIFile(filename: "test.c", directory: "/some/dir")
> !2 = !{i32 7, !"Dwarf Version", i32 5}
> !3 = !{i32 2, !"Debug Info Version", i32 3}
> ```

eddyz87 commented:
> > define void @buz() #0 {
> > entry:
> >   tail call void asm sideeffect "r0 = r1; exit;", ""()
> >   unreachable
> > }
> > ```
> 
> As suggested by @nikic, I will later have a patch to prevent 'unreachable' for naked functions for all architectures. So this test will not be included.

As far as I understand:
- that would be a separate PR merged at some later point;
- the code change in `BPFDAGToDAGISel::Select` is only necessary to handle naked functions.

Thus, I'd keep the test for now, just to have all code paths covered.

yonghong-song commented:
> > > define void @buz() #0 {
> > > entry:
> > > tail call void asm sideeffect "r0 = r1; exit;", ""()
> > > unreachable
> > > }
> > 
> > 
> > As suggested by @nikic, I will later have a patch to prevent 'unreachable' for naked functions for all architectures. So this test will not be included.
> 
> As far as I understand:
> 
>     * that would be a separate PR merged at some later point;
Yes.
> 
>     * the code change in `BPFDAGToDAGISel::Select` is only necessary to handle naked functions.
Yes.
> 
> 
> Thus, I'd keep the test for now, just to have all code paths covered.
Okay, will keep the test for now at least.

inclyc commented on llvm/lib/Target/BPF/BPFISelLowering.cpp:
```diff
@@ -521,10 +526,12 @@ SDValue BPFTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
-    fail(CLI.DL, DAG,
-         Twine("A call to built-in function '" + StringRef(E->getSymbol()) +
-               "' is not supported."));
+    if (strcmp(E->getSymbol(), BPF_UNREACHABLE) != 0) {
```
```suggestion
    if (StringRef(E->getSymbol()) != BPF_UNREACHABLE) {
```

inclyc reviewed (commented):
I can do a quick search on GitHub and find existing examples that generate `__bpf_unreachable();`. Could the `call` we generate conflict with existing ones / user defined ones?

eddyz87 commented on llvm/lib/Target/BPF/BPF.h:
```diff
@@ -22,6 +22,8 @@ class BPFTargetMachine;
 class InstructionSelector;
 class PassRegistry;
 
+static const char *BPF_UNREACHABLE = "bpf_unreachable";
```
Q: given that reserved identifiers start from `__`, should we use `__bpf_unreachable` here (and on kernel side)?
Otherwise this is in conflict with section 7.1.3 of [here](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf).

eddyz87 commented on llvm/lib/Target/BPF/BTFDebug.cpp:
```diff
@@ -1622,6 +1627,25 @@ void BTFDebug::endModule() {
   // Collect global types/variables except MapDef globals.
   processGlobals(false);
 
+  // Create a BTF entry for func BPF_UNREACHABLE.
```
Special cases added in `BTFDebug.cpp` can be avoided with the attached patch:
[bpf-unreachable.patch.txt](https://github.com/user-attachments/files/20378501/bpf-unreachable.patch.txt)
The idea is to generate proper DISubprogram for the `bpf_unreachable` as for kfunc, so that BTFDebug logic works as-is.

yonghong-song replied:
I looked at the patch. Generally looks better as it removed a couple "hack" in BTFDebug.cpp. I would prefer to generate DISubprogram unless M->debug_compile_units() is non-NULL, that is, dwarf is already enabled. I will make necessary adjustment based on your code.

eddyz87 commented on llvm/lib/Target/BPF/BPFMIPeephole.cpp:
```diff
@@ -320,6 +320,7 @@ struct BPFMIPreEmitPeephole : public MachineFunctionPass {
   bool adjustBranch();
   bool insertMissingCallerSavedSpills();
   bool removeMayGotoZero();
+  bool addExitAfterUnreachable();
```
Unexpectedly, generating `exit` at selection DAG level is a head scratcher. The following seems to work:

```cpp
@@ -759,7 +774,10 @@ SDValue BPFTargetLowering::LowerTRAP(SDValue Op, SelectionDAG &DAG) const {
   CLI.DL = DL;
   CLI.NoMerge = false;
   CLI.DoesNotReturn = true;
-  return LowerCall(CLI, InVals);
+  SDValue Chain = LowerCall(CLI, InVals);
+  SDValue Glue = Chain.getValue(1);
+  return DAG.getNode(BPFISD::RET_GLUE, DL, MVT::Other,
+                     SmallVector<SDValue>{Chain, Glue});
 }
```

Relies on `BPFretglue` translation to `exit` as described in `BPFInstrInfo.td`.

yonghong-song replied:
The above seems working but it will add 'exit' after __every__ bpf_unreachable() calls. For example, in IR, if we have
```
<...>
unreachable
<...>
unreachable
<...>
unreachable
```

The above code will generate
```
<...>
bpf_unreachable()
exit
<...>
bpf_unreachable()
exit
<...>
bpf_unreachable()
exit
```

I would like to have
```
<...>
bpf_unreachable()
<...>
bpf_unreachable()
<...>
bpf_unreachable()
exit
```

I will be a trivial change on top of my current patch:
```
diff --git a/llvm/lib/Target/BPF/BPFMIPeephole.cpp b/llvm/lib/Target/BPF/BPFMIPeephole.cpp
index a949d98e6de5..bf8188cc2528 100644
--- a/llvm/lib/Target/BPF/BPFMIPeephole.cpp
+++ b/llvm/lib/Target/BPF/BPFMIPeephole.cpp
@@ -742,8 +742,8 @@ bool BPFMIPreEmitPeephole::removeMayGotoZero() {
 bool BPFMIPreEmitPeephole::addExitAfterUnreachable() {
   MachineBasicBlock &MBB = MF->back();
   MachineInstr &MI = MBB.back();
-  if (MI.getOpcode() != BPF::JAL || !MI.getOperand(0).isSymbol() ||
-      strcmp(MI.getOperand(0).getSymbolName(), BPF_UNREACHABLE))
+  if (MI.getOpcode() != BPF::JAL || !MI.getOperand(0).isGlobal() ||
+      MI.getOperand(0).getGlobal()->getName() != BPF_UNREACHABLE)
     return false;
 
   BuildMI(&MBB, MI.getDebugLoc(), TII->get(BPF::RET));
```

eddyz87 replied:
I see, you mean if it is not the last instruction.
Tbh, I don't think this call would ever happen as non-last instruction: `unreachable` is a terminator. Anyways, this is a minor detail, just thought that keeping everything in one place might be simpler.

yonghong-song replied:
I have the same feeling as well about 'unreachable' IR will be *always* the last insn/terminator. But who knows, during kernel code development, Alexei and I discussed and concluded that we should handle the possibility that 'unreachable' maybe in the middle of IR.

BTW, just updated a new version to incorporate some of your changes. Please take a look!

eddyz87 commented:
> Okay, with __bpf_unreachable, we will have a conflict with kernel. In bpf_helpers.h,
> 
> ```
> /*
>  * Helper macro to throw a compilation error if __bpf_unreachable() gets
>  * built into the resulting code. This works given BPF back end does not
>  * implement __builtin_trap(). This is useful to assert that certain paths
>  * of the program code are never used and hence eliminated by the compiler.
>  *
>  * For example, consider a switch statement that covers known cases used by
>  * the program. __bpf_unreachable() can then reside in the default case. If
>  * the program gets extended such that a case is not covered in the switch
>  * statement, then it will throw a build error due to the default case not
>  * being compiled out.
>  */
> #ifndef __bpf_unreachable
> # define __bpf_unreachable()    __builtin_trap()
> #endif
> ```
> 
> We need to pick a different symbol name.

`__builtin_trap()` is now supported by the backend and is translated as a call to `__bpf_unreachable`, e.g.:

```
$ cat ~/work/tmp/test_trap.c 
void foo(void);
void bar(void);
void buz(int p) {
  if (p) {
    foo();
    __builtin_trap();
  } else {
    bar();
  }
}
$ build/bin/clang --target=bpf -mcpu=v3 -S -o - ~/work/tmp/test_trap.c 
	...
buz:                                    # @buz
# %bb.0:                                # %entry
	*(u32 *)(r10 - 4) = w1
	w1 = *(u32 *)(r10 - 4)
	if w1 == 0 goto LBB0_2
	goto LBB0_1
LBB0_1:                                 # %if.then
	call foo
	call __bpf_unreachable
	goto LBB0_3
LBB0_2:                                 # %if.else
	call bar
	goto LBB0_3
LBB0_3:                                 # %if.end
	exit
.Lfunc_end0:
	...
```

Looks like `bpf_helpers.h` now needs to distinguish between compile time and load time asserts (use different macro definitions).

eddyz87 commented:
@yonghong-song ,

> BTW, just updated a new version to incorporate some of your changes. Please take a look!

The changes lgtm, I also tried end-to-end by adjusting your kernel series to `__bpf_unreachable` and inserting `__builtin_trap()` call in one of the C tests. Verifier reports error as expected.

efriedma-quic commented:
I'm still not completely happy with this; see #136603 for my most recent commentary on producing warnings based on compiler optimizations.  (That's not BPF, but under the proposal here, a similar construct compiled to BPF would produce a BPF verifier error.)  But Linux kernel developers have a different perspective on how a compiler should work... I'm not sure how to bridge the gap.

4ast commented:
> I'm still not completely happy with this; see #136603 for my most recent commentary on producing warnings based on compiler optimizations. (That's not BPF, but under the proposal here, a similar construct compiled to BPF would produce a BPF verifier error.) But Linux kernel developers have a different perspective on how a compiler should work... I'm not sure how to bridge the gap.

I don't see how issues with __builtin_dynamic_object_size() related to this diff.

efriedma-quic commented:
> I don't see how issues with __builtin_dynamic_object_size() related to this diff.

It's not really about __builtin_dynamic_object_size; it's about the way unrolling works.

4ast commented:
> > I don't see how issues with __builtin_dynamic_object_size() related to this diff.
> 
> It's not really about __builtin_dynamic_object_size; it's about the way unrolling works.

and? This patch has nothing to do with unrolling. are you commenting on the right diff?

yonghong-song commented:
Just updated another change which should resolve kernel build issue. At the same time, I will investigate to do different handling of 'unreachable' and '@llvm.trap'.

yonghong-song commented:
sorry, may need another change (basically trying to make -ftrap-function change as later as possible in middle end optimization).

Commit:
Fix issues in BTFDebug for potential missing __bpf_trap btf

The funciton __bpf_trap is created during selectiondag lowering.
But the function usage could be removed during later machine-level
optimization. In such cases, we will generate btf for __bpf_trap
in endModule().

yonghong-song commented:
Rebase on top of latest llvm-project code base.

RKSimon commented on llvm/lib/Target/BPF/BPF.h:
```diff
@@ -22,6 +22,8 @@ class BPFTargetMachine;
 class InstructionSelector;
 class PassRegistry;
 
+static const char *BPF_TRAP = "__bpf_trap";
```
@yonghong-song I'm seeing a lot of gcc warnings: 
```
In file included from /home/simon/LLVM/llvm-project/llvm/lib/Target/BPF/BPFMISimplifyPatchable.cpp:30:
/home/simon/LLVM/llvm-project/llvm/lib/Target/BPF/BPF.h:25:20: warning: ‘llvm::BPF_TRAP’ defined but not used [-Wunused-variable]
   25 | static const char *BPF_TRAP = "__bpf_trap";
```
Any chance you can change this to:
```c
#define BPF_TRAP "__bpf_trap"
```

yonghong-song replied:
@RKSimon I could do the change. But can you let me know how to reproduce the warning? Maybe I missed some compilation flags?
